### ISSUE TITLE
Error on webhook stage when waiting for completion

### ISSUE THREAD

**Issue by dnorrell** [spinnaker/spinnaker#4881]
### Issue Summary:

Webhook stage fails while waiting for completion - error parsing response

### Cloud Provider(s):

GCP

### Environment:

Spinnaker 1.16.0 local install on Ubuntu 16.04

### Feature Area:

orca/webhook stage

### Description:

Setup a webhook stage using 'wait for completion'. The stage starts but when the webhook is monitored it failed. This was working before upgrading to 1.16.0

### Steps to Reproduce:

1. Setup webhook stage + wait for completion, using status from location header
2. Run pipeline
3. Stage fires but the monitor fails with the following error

### Additional Details:

UI shows:
```
Exception ( Monitor Webhook )
Unable to map context to class com.netflix.spinnaker.orca.webhook.pipeline.WebhookStage$StageData
```

Logs:

com.fasterxml.jackson.databind.exc.MismatchedInputException: Cannot construct instance of 'java.util.LinkedHashMap' (although at least one Creator exists): no String-argument constructor/factory method to deserialize from String value ('{\"statusMessage\":\"RUNNING\"}\n')\n at [Source: UNKNOWN; line: -1, column: -1] (through reference chain: com.netflix.spinnaker.orca.webhook.pipeline.WebhookStage$StageData[\"webhook\"]->com.netflix.spinnaker.orca.webhook.pipeline.WebhookStage$WebhookResponseStageData[\"monitor\"]->com.netflix.spinnaker.orca.webhook.pipeline.WebhookStage$WebhookMonitorResponseStageData[\"body\"])\n\tat com.fasterxml.jackson.databind.exc.MismatchedInputException.from(MismatchedInputException.java:63)\n\tat com.fasterxml.jackson.databind.DeserializationContext.reportInputMismatch(DeserializationContext.java:1343)\n\tat com.fasterxml.jackson.databind.DeserializationContext.handleMissingInstantiator(DeserializationContext.java:1032)\n\tat com.fasterxml.jackson.databind.deser.ValueInstantiator._createFromStringFallbacks(ValueInstantiator.java:371)\n\tat com.fasterxml.jackson.databind.deser.std.StdValueInstantiator.createFromString(StdValueInstantiator.java:323)\n\tat com.fasterxml.jackson.databind.deser.std.MapDeserializer.deserialize(MapDeserializer.java:357)\n\tat com.fasterxml.jackson.databind.deser.std.MapDeserializer.deserialize(MapDeserializer.java:29)\n\tat com.fasterxml.jackson.databind.deser.impl.MethodProperty.deserializeAndSet(MethodProperty.java:129)\n\tat com.fasterxml.jackson.databind.deser.BeanDeserializer.vanillaDeserialize(BeanDeserializer.java:288)\n\tat com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:151)\n\tat com.fasterxml.jackson.databind.deser.impl.MethodProperty.deserializeAndSet(MethodProperty.java:129)\n\tat com.fasterxml.jackson.databind.deser.BeanDeserializer.vanillaDeserialize(BeanDeserializer.java:288)\n\tat com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:151)\n\tat com.fasterxml.jackson.databind.deser.impl.MethodProperty.deserializeAndSet(MethodProperty.java:129)\n\tat com.fasterxml.jackson.databind.deser.BeanDeserializer.vanillaDeserialize(BeanDeserializer.java:288)\n\tat com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:151)\n\tat com.fasterxml.jackson.databind.ObjectMapper._readValue(ObjectMapper.java:3984)\n\tat com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:2276)\n\tat com.netflix.spinnaker.orca.pipeline.model.Stage.mapTo(Stage.java:616)\n\t... 31 more\n",

**Comment by gal-yardeni** [comment]
Hi @dnorrell , we tried to reproduce the issue but it worked as expected.
Can you please provide us more info about the request you are trying to do? You can go to the webhook configuration and paste here the json payload (click on "edit stage as json") we can take a closer look. 
This is the payload we tried and worked - the /getRedirect contains a location header that returns "message": "Success".
{
  "customHeaders": {},
  "method": "GET",
  "name": "Webhook",
  "statusJsonPath": "$.message",
  "statusUrlResolution": "locationHeader",
  "successStatuses": "Success",
  "type": "webhook",
  "url": "http:/<myapp>/goRedirect",
  "waitForCompletion": true
}

**Comment by dnorrell** [comment]
The stage config is:
```
{
  "canceledStatuses": "CANCELLED",
  "completeOtherBranchesThenFail": false,
  "continuePipeline": true,
  "failPipeline": false,
  "method": "POST",
  "name": "Deploy to xxxxxxx",
  "payload": {
    "config": "${ config }",
    "environment": "${ environment }",
    "image": "${ image }"
  },
  "statusJsonPath": "$.statusMessage",
  "statusUrlResolution": "locationHeader",
  "successStatuses": "SUCCESS",
  "terminalStatuses": "TERMINAL",
  "type": "webhook",
  "url": "https://xxxxxxxxxxxxx/deploy",
  "waitForCompletion": true
}
```
The response payload from the webhook status URL is:
```
{"statusMessage":"RUNNING","progressMessage":"","error":""}
```

**Comment by dnorrell** [comment]
As I said, I had this working on 1.15.x but it seemed to break in 1.16.

**Comment by gal-yardeni** [comment]
@dnorrell Tried to reproduce it with different endpoints and payloads and I don't seem to hit this issue. 
Can you share the response coming from the locationHeader endpoint? Looking at the logs you shared, it looks like it's maybe not a valid json (('{"statusMessage":"RUNNING"}\n')\).
Do you see this issue every time you're trying to run the webhook pipeline? Any other info you can provide will be useful.

**Comment by eboumendil** [comment]
I have the same issue happening since I upgraded to v1.16.
Here are the logs:
```
2019-10-01 20:53:35.880  WARN 1 --- [    handlers-18] c.n.s.o.e.DefaultExceptionHandler        : [] Error occurred during task monitorWebhook
java.lang.IllegalArgumentException: Unable to map context to class com.netflix.spinnaker.orca.webhook.pipeline.WebhookStage$StageData
	at com.netflix.spinnaker.orca.pipeline.model.Stage.mapTo(Stage.java:621)
	at com.netflix.spinnaker.orca.pipeline.model.Stage.mapTo(Stage.java:604)
	at com.netflix.spinnaker.orca.pipeline.model.Stage$mapTo.call(Unknown Source)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:47)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:115)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:127)
	at com.netflix.spinnaker.orca.webhook.tasks.MonitorWebhookTask.execute(MonitorWebhookTask.groovy:64)
	at com.netflix.spinnaker.orca.q.handler.RunTaskHandler$handle$1$2$1.invoke(RunTaskHandler.kt:113)
	at com.netflix.spinnaker.orca.q.handler.RunTaskHandler$handle$1$2$1.invoke(RunTaskHandler.kt:69)
	at com.netflix.spinnaker.orca.q.handler.RunTaskHandler.withLoggingContext(RunTaskHandler.kt:318)
	at com.netflix.spinnaker.orca.q.handler.RunTaskHandler.access$withLoggingContext(RunTaskHandler.kt:69)
	at com.netflix.spinnaker.orca.q.handler.RunTaskHandler$handle$1$2.invoke(RunTaskHandler.kt:112)
	at com.netflix.spinnaker.orca.q.handler.RunTaskHandler$handle$1$2.invoke(RunTaskHandler.kt:69)
	at com.netflix.spinnaker.orca.q.handler.AuthenticationAware$sam$java_util_concurrent_Callable$0.call(AuthenticationAware.kt)
	at com.netflix.spinnaker.security.AuthenticatedRequest.lambda$propagate$0(AuthenticatedRequest.java:127)
	at com.netflix.spinnaker.orca.q.handler.AuthenticationAware$DefaultImpls.withAuth(AuthenticationAware.kt:51)
	at com.netflix.spinnaker.orca.q.handler.RunTaskHandler.withAuth(RunTaskHandler.kt:69)
	at com.netflix.spinnaker.orca.q.handler.RunTaskHandler$handle$1.invoke(RunTaskHandler.kt:111)
	at com.netflix.spinnaker.orca.q.handler.RunTaskHandler$handle$1.invoke(RunTaskHandler.kt:69)
	at com.netflix.spinnaker.orca.q.handler.RunTaskHandler$withTask$1.invoke(RunTaskHandler.kt:196)
	at com.netflix.spinnaker.orca.q.handler.RunTaskHandler$withTask$1.invoke(RunTaskHandler.kt:69)
	at com.netflix.spinnaker.orca.q.handler.OrcaMessageHandler$withTask$1.invoke(OrcaMessageHandler.kt:57)
	at com.netflix.spinnaker.orca.q.handler.OrcaMessageHandler$withTask$1.invoke(OrcaMessageHandler.kt:41)
	at com.netflix.spinnaker.orca.q.handler.OrcaMessageHandler$withStage$1.invoke(OrcaMessageHandler.kt:67)
	at com.netflix.spinnaker.orca.q.handler.OrcaMessageHandler$withStage$1.invoke(OrcaMessageHandler.kt:41)
	at com.netflix.spinnaker.orca.q.handler.OrcaMessageHandler$DefaultImpls.withExecution(OrcaMessageHandler.kt:76)
	at com.netflix.spinnaker.orca.q.handler.RunTaskHandler.withExecution(RunTaskHandler.kt:69)
	at com.netflix.spinnaker.orca.q.handler.OrcaMessageHandler$DefaultImpls.withStage(OrcaMessageHandler.kt:63)
	at com.netflix.spinnaker.orca.q.handler.RunTaskHandler.withStage(RunTaskHandler.kt:69)
	at com.netflix.spinnaker.orca.q.handler.OrcaMessageHandler$DefaultImpls.withTask(OrcaMessageHandler.kt:50)
	at com.netflix.spinnaker.orca.q.handler.RunTaskHandler.withTask(RunTaskHandler.kt:69)
	at com.netflix.spinnaker.orca.q.handler.RunTaskHandler.withTask(RunTaskHandler.kt:189)
	at com.netflix.spinnaker.orca.q.handler.RunTaskHandler.handle(RunTaskHandler.kt:82)
	at com.netflix.spinnaker.orca.q.handler.RunTaskHandler.handle(RunTaskHandler.kt:69)
	at com.netflix.spinnaker.q.MessageHandler$DefaultImpls.invoke(MessageHandler.kt:36)
	at com.netflix.spinnaker.orca.q.handler.OrcaMessageHandler$DefaultImpls.invoke(OrcaMessageHandler.kt)
	at com.netflix.spinnaker.orca.q.handler.RunTaskHandler.invoke(RunTaskHandler.kt:69)
	at com.netflix.spinnaker.orca.q.audit.ExecutionTrackingMessageHandlerPostProcessor$ExecutionTrackingMessageHandlerProxy.invoke(ExecutionTrackingMessageHandlerPostProcessor.kt:69)
	at com.netflix.spinnaker.q.QueueProcessor$pollOnce$1$1.run(QueueProcessor.kt:82)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: com.fasterxml.jackson.databind.exc.MismatchedInputException: Cannot construct instance of `java.util.LinkedHashMap` (although at least one Creator exists): no String-argument constructor/factory method to deserialize from String value ('{"ok":true,"error":"","error_raw":null}
')
 at [Source: UNKNOWN; line: -1, column: -1] (through reference chain: com.netflix.spinnaker.orca.webhook.pipeline.WebhookStage$StageData["webhook"]->com.netflix.spinnaker.orca.webhook.pipeline.WebhookStage$WebhookResponseStageData["body"])
	at com.fasterxml.jackson.databind.exc.MismatchedInputException.from(MismatchedInputException.java:63)
	at com.fasterxml.jackson.databind.DeserializationContext.reportInputMismatch(DeserializationContext.java:1343)
	at com.fasterxml.jackson.databind.DeserializationContext.handleMissingInstantiator(DeserializationContext.java:1032)
	at com.fasterxml.jackson.databind.deser.ValueInstantiator._createFromStringFallbacks(ValueInstantiator.java:371)
	at com.fasterxml.jackson.databind.deser.std.StdValueInstantiator.createFromString(StdValueInstantiator.java:323)
	at com.fasterxml.jackson.databind.deser.std.MapDeserializer.deserialize(MapDeserializer.java:357)
	at com.fasterxml.jackson.databind.deser.std.MapDeserializer.deserialize(MapDeserializer.java:29)
	at com.fasterxml.jackson.databind.deser.impl.MethodProperty.deserializeAndSet(MethodProperty.java:129)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.vanillaDeserialize(BeanDeserializer.java:288)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:151)
	at com.fasterxml.jackson.databind.deser.impl.MethodProperty.deserializeAndSet(MethodProperty.java:129)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.vanillaDeserialize(BeanDeserializer.java:288)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:151)
	at com.fasterxml.jackson.databind.ObjectMapper._readValue(ObjectMapper.java:3984)
	at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:2276)
	at com.netflix.spinnaker.orca.pipeline.model.Stage.mapTo(Stage.java:616)
	... 41 common frames omitted
2019-10-01 20:53:35.880 ERROR 1 --- [    handlers-18] c.n.s.orca.q.handler.RunTaskHandler      : [] Error running MonitorWebhookTask for pipeline[01DP4HK0R732CTG4359ARSJWR9]
java.lang.IllegalArgumentException: Unable to map context to class com.netflix.spinnaker.orca.webhook.pipeline.WebhookStage$StageData
	at com.netflix.spinnaker.orca.pipeline.model.Stage.mapTo(Stage.java:621)
	at com.netflix.spinnaker.orca.pipeline.model.Stage.mapTo(Stage.java:604)
	at com.netflix.spinnaker.orca.pipeline.model.Stage$mapTo.call(Unknown Source)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:47)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:115)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:127)
	at com.netflix.spinnaker.orca.webhook.tasks.MonitorWebhookTask.execute(MonitorWebhookTask.groovy:64)
	at com.netflix.spinnaker.orca.q.handler.RunTaskHandler$handle$1$2$1.invoke(RunTaskHandler.kt:113)
	at com.netflix.spinnaker.orca.q.handler.RunTaskHandler$handle$1$2$1.invoke(RunTaskHandler.kt:69)
	at com.netflix.spinnaker.orca.q.handler.RunTaskHandler.withLoggingContext(RunTaskHandler.kt:318)
	at com.netflix.spinnaker.orca.q.handler.RunTaskHandler.access$withLoggingContext(RunTaskHandler.kt:69)
	at com.netflix.spinnaker.orca.q.handler.RunTaskHandler$handle$1$2.invoke(RunTaskHandler.kt:112)
	at com.netflix.spinnaker.orca.q.handler.RunTaskHandler$handle$1$2.invoke(RunTaskHandler.kt:69)
	at com.netflix.spinnaker.orca.q.handler.AuthenticationAware$sam$java_util_concurrent_Callable$0.call(AuthenticationAware.kt)
	at com.netflix.spinnaker.security.AuthenticatedRequest.lambda$propagate$0(AuthenticatedRequest.java:127)
	at com.netflix.spinnaker.orca.q.handler.AuthenticationAware$DefaultImpls.withAuth(AuthenticationAware.kt:51)
	at com.netflix.spinnaker.orca.q.handler.RunTaskHandler.withAuth(RunTaskHandler.kt:69)
	at com.netflix.spinnaker.orca.q.handler.RunTaskHandler$handle$1.invoke(RunTaskHandler.kt:111)
	at com.netflix.spinnaker.orca.q.handler.RunTaskHandler$handle$1.invoke(RunTaskHandler.kt:69)
	at com.netflix.spinnaker.orca.q.handler.RunTaskHandler$withTask$1.invoke(RunTaskHandler.kt:196)
	at com.netflix.spinnaker.orca.q.handler.RunTaskHandler$withTask$1.invoke(RunTaskHandler.kt:69)
	at com.netflix.spinnaker.orca.q.handler.OrcaMessageHandler$withTask$1.invoke(OrcaMessageHandler.kt:57)
	at com.netflix.spinnaker.orca.q.handler.OrcaMessageHandler$withTask$1.invoke(OrcaMessageHandler.kt:41)
	at com.netflix.spinnaker.orca.q.handler.OrcaMessageHandler$withStage$1.invoke(OrcaMessageHandler.kt:67)
	at com.netflix.spinnaker.orca.q.handler.OrcaMessageHandler$withStage$1.invoke(OrcaMessageHandler.kt:41)
	at com.netflix.spinnaker.orca.q.handler.OrcaMessageHandler$DefaultImpls.withExecution(OrcaMessageHandler.kt:76)
	at com.netflix.spinnaker.orca.q.handler.RunTaskHandler.withExecution(RunTaskHandler.kt:69)
	at com.netflix.spinnaker.orca.q.handler.OrcaMessageHandler$DefaultImpls.withStage(OrcaMessageHandler.kt:63)
	at com.netflix.spinnaker.orca.q.handler.RunTaskHandler.withStage(RunTaskHandler.kt:69)
	at com.netflix.spinnaker.orca.q.handler.OrcaMessageHandler$DefaultImpls.withTask(OrcaMessageHandler.kt:50)
	at com.netflix.spinnaker.orca.q.handler.RunTaskHandler.withTask(RunTaskHandler.kt:69)
	at com.netflix.spinnaker.orca.q.handler.RunTaskHandler.withTask(RunTaskHandler.kt:189)
	at com.netflix.spinnaker.orca.q.handler.RunTaskHandler.handle(RunTaskHandler.kt:82)
	at com.netflix.spinnaker.orca.q.handler.RunTaskHandler.handle(RunTaskHandler.kt:69)
	at com.netflix.spinnaker.q.MessageHandler$DefaultImpls.invoke(MessageHandler.kt:36)
	at com.netflix.spinnaker.orca.q.handler.OrcaMessageHandler$DefaultImpls.invoke(OrcaMessageHandler.kt)
	at com.netflix.spinnaker.orca.q.handler.RunTaskHandler.invoke(RunTaskHandler.kt:69)
	at com.netflix.spinnaker.orca.q.audit.ExecutionTrackingMessageHandlerPostProcessor$ExecutionTrackingMessageHandlerProxy.invoke(ExecutionTrackingMessageHandlerPostProcessor.kt:69)
	at com.netflix.spinnaker.q.QueueProcessor$pollOnce$1$1.run(QueueProcessor.kt:82)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: com.fasterxml.jackson.databind.exc.MismatchedInputException: Cannot construct instance of `java.util.LinkedHashMap` (although at least one Creator exists): no String-argument constructor/factory method to deserialize from String value ('{"ok":true,"error":"","error_raw":null}
')
 at [Source: UNKNOWN; line: -1, column: -1] (through reference chain: com.netflix.spinnaker.orca.webhook.pipeline.WebhookStage$StageData["webhook"]->com.netflix.spinnaker.orca.webhook.pipeline.WebhookStage$WebhookResponseStageData["body"])
	at com.fasterxml.jackson.databind.exc.MismatchedInputException.from(MismatchedInputException.java:63)
	at com.fasterxml.jackson.databind.DeserializationContext.reportInputMismatch(DeserializationContext.java:1343)
	at com.fasterxml.jackson.databind.DeserializationContext.handleMissingInstantiator(DeserializationContext.java:1032)
	at com.fasterxml.jackson.databind.deser.ValueInstantiator._createFromStringFallbacks(ValueInstantiator.java:371)
	at com.fasterxml.jackson.databind.deser.std.StdValueInstantiator.createFromString(StdValueInstantiator.java:323)
	at com.fasterxml.jackson.databind.deser.std.MapDeserializer.deserialize(MapDeserializer.java:357)
	at com.fasterxml.jackson.databind.deser.std.MapDeserializer.deserialize(MapDeserializer.java:29)
	at com.fasterxml.jackson.databind.deser.impl.MethodProperty.deserializeAndSet(MethodProperty.java:129)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.vanillaDeserialize(BeanDeserializer.java:288)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:151)
	at com.fasterxml.jackson.databind.deser.impl.MethodProperty.deserializeAndSet(MethodProperty.java:129)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.vanillaDeserialize(BeanDeserializer.java:288)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:151)
	at com.fasterxml.jackson.databind.ObjectMapper._readValue(ObjectMapper.java:3984)
	at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:2276)
	at com.netflix.spinnaker.orca.pipeline.model.Stage.mapTo(Stage.java:616)
	... 41 common frames omitted
2019-10-01 20:53:35.933  INFO 1 --- [    scheduler-3] com.netflix.spinnaker.q.QueueProcessor   : [] Received message CancelStage(executionType=pipeline, executionId=01DP4HK0R732CTG4359ARSJWR9, application=houston, stageId=01DP4HK0VVQM80D8K740CGSCMR)
```

**Comment by marchello2000** [comment]
merged, should make it to 1.16.3 release

**Comment by marchello2000** [comment]
@dnorrell & @eboumendil thanks  for reporting and sorry it took us a while to figure this one out
